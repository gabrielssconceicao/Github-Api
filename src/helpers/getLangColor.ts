export const languageColors: Record<string, string> = {
  JavaScript: '#f7df1e',
  Python: '#3572A5',
  Java: '#b07219',
  'C++': '#f34b7d',
  C: '#555555',
  'C#': '#178600',
  PHP: '#4F5D95',
  TypeScript: '#2b7489',
  Swift: '#ffac45',
  Ruby: '#CC342D',
  Go: '#00ADD8',
  R: '#198CE7',
  Kotlin: '#F18E33',
  Scala: '#c22d40',
  Rust: '#dea584',
  Dart: '#00B4AB',
  Perl: '#0298c3',
  Lua: '#000080',
  Matlab: '#e16737',
  Haskell: '#5e5086',
  SQL: '#e38c00',
  HTML: '#e34c26',
  CSS: '#563d7c',
  'Objective-C': '#438eff',
  Assembly: '#6E4C13',
  Shell: '#89e051',
  PowerShell: '#012456',
  Groovy: '#e69f56',
  Vue: '#41B883',
  Dockerfile: '#384d54',
  Clojure: '#db5855',
  Elixir: '#6e4a7e',
  D: '#ba595e',
  Julia: '#a270ba',
  Arduino: '#bd79d1',
  Verilog: '#b2b7f8',
  VHDL: '#adb2cb',
  Ada: '#02f88c',
  Scheme: '#1e4aec',
  'Objective-C++': '#6866fb',
  Crystal: '#000100',
};
export const getLangColor = (lang: string = ''): string => {
  let language = '#ffffff';
  if (lang in languageColors) {
    language = languageColors[lang];
  }
  return language;
};
