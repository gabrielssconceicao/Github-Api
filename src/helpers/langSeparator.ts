import type { IRepos } from '@/types/repos';
type ILang = { name: string; count: number };
export const langSeparator = (repos: IRepos[]): ILang[] => {
  const langs: ILang[] = [];
  repos.forEach((repo) => {
    const lang = repo.language;
    const index = langs.findIndex((l) => l.name === lang);
    console.log(index);
    if (index === -1) {
      langs.push({ name: lang, count: 1 });
    } else {
      langs[index].count += 1;
    }
    console.log(langs);
  });
  return langs;
};
