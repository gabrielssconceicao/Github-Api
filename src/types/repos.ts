export interface IRepos {
  name: string;
  description: string | null;
  html_url: string;
  language: string;
}
