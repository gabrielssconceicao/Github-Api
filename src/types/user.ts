export interface IUser {
  name: null | string;
  login: string;
  avatar_url: string;
  followers: number;
  following: number;
  location: null | string;
  html_url: string;
  bio: null | string;
  public_repos: number;
  repos_url: string;
}

export interface IUserResponse {
  user: IUser;
  status: number;
}
