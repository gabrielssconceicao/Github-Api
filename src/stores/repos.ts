import { defineStore } from 'pinia';
import { ref, computed } from 'vue';

import { type IRepos } from '@/types/repos';

import { langSeparator } from '@/helpers/langSeparator';
import { useLoadStore } from './load';
export const useReposStore = defineStore('repos', () => {
  const repos = ref<IRepos[]>([]);
  const reposUrl = ref('');
  const filterValue = ref('');

  function setRepos(reposValue: IRepos[]) {
    repos.value = reposValue;
  }

  const setFilter = (value: string) => {
    filterValue.value = value;
  };

  function setReposUrl(url: string) {
    reposUrl.value = url;
  }
  async function getRepos() {
    const useLoad = useLoadStore();
    useLoad.setLoading();
    try {
      const response = await fetch(reposUrl.value);
      const data = await response.json();
      setRepos(data);
    } catch (error) {
      setRepos([]);
    } finally {
      useLoad.setNotLoading();
    }
  }

  const filteredRepos = computed(() => {
    return filterValue.value !== ''
      ? repos.value.filter((repo) => repo.language === filterValue.value)
      : repos.value;
  });

  const filteredReposCount = computed(
    () => repos.value && langSeparator(repos.value),
  );

  return {
    filteredRepos,
    getRepos,
    filteredReposCount,
    setFilter,
    setReposUrl,
    setRepos,
    repos,
    reposUrl,
    filterValue,
  };
});
