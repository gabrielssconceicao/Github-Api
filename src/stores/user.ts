import { ref, computed } from 'vue';
import { defineStore } from 'pinia';
import type { IUserResponse } from '@/types/user';
import { fetchUser } from '@/hooks/fetchUser';
import { useReposStore } from './repos';
export const useUserStore = defineStore('user', () => {
  const githubUser = ref<IUserResponse>({} as IUserResponse);
  const inputValue = ref('');
  async function searchUser() {
    const reposStore = useReposStore();
    const data = await fetchUser(inputValue.value);
    githubUser.value = data;
    reposStore.setReposUrl(data.user.repos_url);
  }

  function setInputValue(value: string) {
    inputValue.value = value;
  }

  const userName = computed(() => {
    return githubUser.value.user?.name || githubUser.value.user?.login;
  });

  return { githubUser, searchUser, setInputValue, inputValue, userName };
});
