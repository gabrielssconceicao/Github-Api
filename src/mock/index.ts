import { type IUser } from '@/types/user';
import { type IRepos } from '@/types/repos';

export const user: IUser = {
  name: 'Teste User',
  login: 'testeuser',
  avatar_url: 'https://avatars.githubusercontent.com/u/10590076?v=4',
  followers: 0,
  following: 5,
  location: 'Brazil',
  html_url: 'https://github.com/testeuser',
  bio: 'lorem ipsum dolor sit amet',
  public_repos: 5,
};
// JS - 3
// HTML - 2
// python - 1
// Ruby - 1
//Java - 1
//C++ - 1
//Typescript - 1
export const repos: IRepos[] = [
  {
    name: 'Teste Repo 1',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Javascript',
  },
  {
    name: 'Teste Repo 2',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'HTML',
  },
  {
    name: 'Teste Repo 3',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Python',
  },
  {
    name: 'Teste Repo 4',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Javascript',
  },
  {
    name: 'Teste Repo 5',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Javascript',
  },
  {
    name: 'Teste Repo 6',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'HTML',
  },
  {
    name: 'Teste Repo 7',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Ruby',
  },
  {
    name: 'Teste Repo 8',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Java',
  },
  {
    name: 'Teste Repo 9',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'C++',
  },
  {
    name: 'Teste Repo 10',
    description: 'lorem ipsum dolor sit amet',
    html_url: 'https://github.com/testeuser/testerepo',
    language: 'Typescript',
  },
];
