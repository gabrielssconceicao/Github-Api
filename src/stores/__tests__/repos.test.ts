import { describe, it, expect } from 'vitest';
import { useReposStore } from '../repos';
import { useUserStore } from '../user';
import { langSeparator } from '@/helpers/langSeparator';
import { repos as mockRepos } from '@/mock/index';
describe('useReposStore', () => {
  it('init with all empty', () => {
    const useRepos = useReposStore();

    expect(useRepos.repos.length).toBe(0);
    expect(useRepos.reposUrl).toBe('');
    expect(useRepos.filterValue).toBe('');
    expect(useRepos.filteredRepos.length).toBe(0);
    expect(useRepos.filteredReposCount.length).toBe(0);
    expect({
      repos: useRepos.repos,
      reposUrl: useRepos.reposUrl,
      filterValue: useRepos.filterValue,
      filteredRepos: useRepos.filteredRepos,
      filteredReposCount: useRepos.filteredReposCount,
    }).toMatchSnapshot();
  });

  it('should change the value of: repos, repoUrl, filteredRepos, filteredReposCount', async () => {
    const useRepos = useReposStore();
    const useUser = useUserStore();
    useUser.setInputValue('resolveTest');
    await useUser.searchUser();
    useRepos.setReposUrl(useUser.githubUser.user.repos_url);
    await useRepos.getRepos();

    expect(useRepos.repos.length).toBe(mockRepos.length);
    expect(useRepos.reposUrl).toBe(useUser.githubUser.user.repos_url);
    expect(useRepos.filteredRepos.length).toBe(mockRepos.length);
    expect(useRepos.filteredReposCount.length).toBe(
      langSeparator(mockRepos).length,
    );

    expect({
      repos: useRepos.repos,
      reposUrl: useRepos.reposUrl,
      filterValue: useRepos.filterValue,
      filteredRepos: useRepos.filteredRepos,
      filteredReposCount: useRepos.filteredReposCount,
    }).toMatchSnapshot();
  });

  it('should filter repos', async () => {
    const useRepos = useReposStore();
    const useUser = useUserStore();
    useUser.setInputValue('resolveTest');
    await useUser.searchUser();
    useRepos.setReposUrl(useUser.githubUser.user.repos_url);
    await useRepos.getRepos();

    useRepos.setFilter('HTML');
    expect(useRepos.filteredRepos.length).toBe(2);

    useRepos.setFilter('');
    expect(useRepos.filteredRepos.length).toBe(mockRepos.length);

    useRepos.setFilter('HTML');
    expect(useRepos.filteredRepos.length).toBe(2);

    expect({
      repos: useRepos.repos,
      reposUrl: useRepos.reposUrl,
      filterValue: useRepos.filterValue,
      filteredRepos: useRepos.filteredRepos,
      filteredReposCount: useRepos.filteredReposCount,
    }).toMatchSnapshot();
  });
});
