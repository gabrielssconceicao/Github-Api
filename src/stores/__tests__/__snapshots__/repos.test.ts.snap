// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useReposStore > init with all empty 1`] = `
{
  "filterValue": "",
  "filteredRepos": [],
  "filteredReposCount": [],
  "repos": [],
  "reposUrl": "",
}
`;

exports[`useReposStore > should change the value of: repos, repoUrl, filteredRepos, filteredReposCount 1`] = `
{
  "filterValue": "",
  "filteredRepos": [
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 1,
      "language": "Javascript",
      "name": "Teste Repo 1",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 2,
      "language": "HTML",
      "name": "Teste Repo 2",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 3,
      "language": "Python",
      "name": "Teste Repo 3",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 4,
      "language": "Javascript",
      "name": "Teste Repo 4",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 5,
      "language": "Javascript",
      "name": "Teste Repo 5",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 6,
      "language": "HTML",
      "name": "Teste Repo 6",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 7,
      "language": "Ruby",
      "name": "Teste Repo 7",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 8,
      "language": "Java",
      "name": "Teste Repo 8",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 9,
      "language": "C++",
      "name": "Teste Repo 9",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 10,
      "language": "Typescript",
      "name": "Teste Repo 10",
    },
  ],
  "filteredReposCount": [
    {
      "count": 3,
      "name": "Javascript",
    },
    {
      "count": 2,
      "name": "HTML",
    },
    {
      "count": 1,
      "name": "Python",
    },
    {
      "count": 1,
      "name": "Ruby",
    },
    {
      "count": 1,
      "name": "Java",
    },
    {
      "count": 1,
      "name": "C++",
    },
    {
      "count": 1,
      "name": "Typescript",
    },
  ],
  "repos": [
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 1,
      "language": "Javascript",
      "name": "Teste Repo 1",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 2,
      "language": "HTML",
      "name": "Teste Repo 2",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 3,
      "language": "Python",
      "name": "Teste Repo 3",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 4,
      "language": "Javascript",
      "name": "Teste Repo 4",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 5,
      "language": "Javascript",
      "name": "Teste Repo 5",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 6,
      "language": "HTML",
      "name": "Teste Repo 6",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 7,
      "language": "Ruby",
      "name": "Teste Repo 7",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 8,
      "language": "Java",
      "name": "Teste Repo 8",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 9,
      "language": "C++",
      "name": "Teste Repo 9",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 10,
      "language": "Typescript",
      "name": "Teste Repo 10",
    },
  ],
  "reposUrl": "https://api.github.com/users/testeuser/repos",
}
`;

exports[`useReposStore > should filter repos 1`] = `
{
  "filterValue": "HTML",
  "filteredRepos": [
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 2,
      "language": "HTML",
      "name": "Teste Repo 2",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 6,
      "language": "HTML",
      "name": "Teste Repo 6",
    },
  ],
  "filteredReposCount": [
    {
      "count": 3,
      "name": "Javascript",
    },
    {
      "count": 2,
      "name": "HTML",
    },
    {
      "count": 1,
      "name": "Python",
    },
    {
      "count": 1,
      "name": "Ruby",
    },
    {
      "count": 1,
      "name": "Java",
    },
    {
      "count": 1,
      "name": "C++",
    },
    {
      "count": 1,
      "name": "Typescript",
    },
  ],
  "repos": [
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 1,
      "language": "Javascript",
      "name": "Teste Repo 1",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 2,
      "language": "HTML",
      "name": "Teste Repo 2",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 3,
      "language": "Python",
      "name": "Teste Repo 3",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 4,
      "language": "Javascript",
      "name": "Teste Repo 4",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 5,
      "language": "Javascript",
      "name": "Teste Repo 5",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 6,
      "language": "HTML",
      "name": "Teste Repo 6",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 7,
      "language": "Ruby",
      "name": "Teste Repo 7",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 8,
      "language": "Java",
      "name": "Teste Repo 8",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 9,
      "language": "C++",
      "name": "Teste Repo 9",
    },
    {
      "description": "lorem ipsum dolor sit amet",
      "html_url": "https://github.com/testeuser/testerepo",
      "id": 10,
      "language": "Typescript",
      "name": "Teste Repo 10",
    },
  ],
  "reposUrl": "https://api.github.com/users/testeuser/repos",
}
`;
